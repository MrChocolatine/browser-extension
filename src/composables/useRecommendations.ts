import useExtensions from '@/composables/useExtensions/useExtensions';
import { ref, watchEffect } from 'vue';
import { asyncComputed } from '@vueuse/core';
import useHTTPSOnly from '@/composables/useHttpsOnly';

type Recommendation = {
  image?: string;
  title: string;
  description: string;
  anchor: string;
  cta: string;
  ctaURL?: string;
};

const { recommendedExtensions } = useExtensions();

const recommendations = ref<Recommendation[]>([]);

const useRecommendations = () => {
  // Add other recommendations here...
  const httpsOnly = asyncComputed(useHTTPSOnly);

  // watchEffect lets us watch multiple properties and trigger side effects
  watchEffect(() => {
    // We need to listen for changes to recommendedExtensions since it is read asynchronously from browser storage
    recommendations.value = recommendedExtensions.value.map((extension) => {
      const cta = extension.installed && !extension.enabled ? 'Enable' : 'Install';

      return {
        image: `/assets/icons/${extension.icon}`,
        title: `${cta} ${extension.name}`,
        description: extension.longDescription,
        ctaURL: extension.addonUrl,
        cta,
        anchor: `/privacy-recommendations#${extension.id}`,
      };
    });

    // We need to listen for httpsOnly updates since it is also read asynchronously
    if (!httpsOnly.value) {
      recommendations.value.push({
        title: 'Set HTTPS-only mode',
        description:
          'Enabling this security enhancing mode provides a guarantee that all of your connections to websites are upgraded to use HTTPS, and warn you if only HTTP unsafe mode is available.',
        anchor: '/privacy-recommendations#https-only',
        cta: 'Set HTTPS-Only mode',
        ctaURL:
          'https://support.mozilla.org/en-US/kb/https-only-prefs#w_enabledisable-https-only-mode',
      });
    } else {
      // Ugly filter function to remove Https-only mode from the array
      recommendations.value = recommendations.value.filter(
        (r) => r.title !== 'Set HTTPS-only mode',
      );
    }
  });

  return { recommendations };
};

export default useRecommendations;
