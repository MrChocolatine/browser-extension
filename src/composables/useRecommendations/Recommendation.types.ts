export interface Recommendation {
  type: 'extension' | 'setting' | 'warning';
  id: string;
  name: string;
  description: string;
  icon?: string;
  iconType?: 'warning' | 'success' | 'info' | 'leak';
  homeUrl?: string;
  warning?: string;
  ctaLabel: 'install' | 'enable' | 'disable' | undefined;
  ctaUrl?: string;
  installed?: boolean;
  enabled?: boolean;
  activated: boolean;
  ignored: boolean;
}
