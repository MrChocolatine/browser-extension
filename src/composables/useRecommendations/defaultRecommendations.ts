import { Recommendation } from './Recommendation.types';

export const extensions: Recommendation[] = [
  {
    type: 'extension',
    id: 'uBlock0@raymondhill.net',
    name: 'uBlock Origin',
    description: `uBlock Origin is not just a free and open-source “ad blocker“, it's a very efficient content blocker consuming minimal resources.`,
    homeUrl: 'https://ublockorigin.com/',
    icon: 'ubo64.png',
    enabled: false,
    installed: false,
    activated: false,
    ignored: false,
    ctaLabel: 'install',
    ctaUrl: 'https://addons.mozilla.org/firefox/addon/ublock-origin/',
  },
];

export const settings: Recommendation[] = [
  {
    type: 'setting',
    id: 'disable-webrtc',
    name: 'Disable webRTC',
    description: `Disable WebRTC to stop websites using that technology from leaking you IP address, even when you're behind a VPN or Tor.`,
    warning:
      'WebRTC is needed by some websites, for example some browser based videoconference tools.',
    homeUrl: 'https://mullvad.net/en/help/webrtc/',
    activated: true,
    ignored: false,
    ctaLabel: 'disable',
  },
  {
    type: 'setting',
    id: 'https-only-mode',
    name: 'Use HTTPS-only mode',
    description:
      'Enabling this security enhancing mode provides a guarantee that all of your connections to websites are upgraded to use HTTPS, and warn you if only HTTP unsafe mode is available.',
    homeUrl:
      'https://support.mozilla.org/en-US/kb/https-only-prefs#w_enabledisable-https-only-mode',
    activated: false,
    ignored: false,
    ctaLabel: undefined,
  },
  {
    type: 'setting',
    id: 'disable-doh',
    name: 'Disable Mullvad DoH (Encrypted DNS)',
    description: `When you're connected to Mullvad, it's better to use the DNS from the same exit server configured.`,
    homeUrl: 'https://mullvad.net/en/help/webrtc/',
    activated: false,
    ignored: false,
    ctaLabel: undefined,
  },
];

export const isRecommended = (id: string) => {
  return defaultRecommendationsIds.includes(id);
};

export const defaultRecommendations = [...settings, ...extensions];

export const defaultRecommendationsIds = defaultRecommendations.map((rec) => rec.id);
