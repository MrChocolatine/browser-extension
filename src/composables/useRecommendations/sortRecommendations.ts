import { Recommendation } from './Recommendation.types';

/**
 * Sort in the following order:
 * 1. to install
 * 2. disabled
 * 3. ignored
 * 4. activated
 * 5. by name
 * @param recommendations
 */
const sortRecommendations = (recommendations: Recommendation[]) => {
  // Make sure we don't mutate the original array!
  return [...recommendations].sort((a, b) => {
    if (a.activated === b.activated) {
      if (a.enabled === b.enabled) {
        if (a.ignored === b.ignored) {
          return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
        } else {
          return a.ignored ? 1 : -1;
        }
      } else {
        return a.enabled ? 1 : -1;
      }
    }
    return a.activated ? 1 : -1;
  });
};

export default sortRecommendations;
