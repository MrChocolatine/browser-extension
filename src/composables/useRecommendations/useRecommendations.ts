import { computed, shallowRef, watch } from 'vue';
import { management } from 'webextension-polyfill';

import useBrowserStorageLocal from '@/composables/useBrowserStorageLocal';
import useHttpsOnly from '@/composables/useHttpsOnly';
import { Recommendation } from '@/composables/useRecommendations/Recommendation.types';
import {
  defaultRecommendations,
  defaultRecommendationsIds,
} from '@/composables/useRecommendations/defaultRecommendations';

const recommendations = useBrowserStorageLocal<Recommendation[]>(
  'recommendations',
  defaultRecommendations,
);

const updateRecConfig = (id: string, modification: Partial<Recommendation>) => {
  recommendations.value = recommendations.value.map((recommendation) =>
    recommendation.id === id ? { ...recommendation, ...modification } : recommendation,
  );
};

const getRecConfigById = (id: string) => {
  return recommendations.value.find((rec) => rec.id === id);
};

const updateHttpsOnly = async () => {
  const httpsOnly = await useHttpsOnly();

  updateRecConfig('https-only-mode', { activated: httpsOnly });
};

const updateSettings = () => {
  updateHttpsOnly();
};

// Update browser extensions recommendations
export const getCurrentUserRecommendations = async () => {
  const installedAddons = (await management.getAll()).filter((extension) =>
    defaultRecommendationsIds.includes(extension.id),
  );
  const installedAddonsIds = installedAddons.map((addons) => addons.id);

  // Create a disabled extensions ID list
  const disabledIDs = installedAddons.filter((addon) => !addon.enabled).map((addons) => addons.id);

  installedAddons.forEach((extension) => {
    const enabled = !disabledIDs.includes(extension.id);
    const installed = installedAddonsIds.includes(extension.id);

    const partialUpdate: Partial<Recommendation> = {
      ctaLabel: installed ? (!enabled ? 'enable' : undefined) : 'install',
      enabled,
      installed,
      activated: installed && enabled,
    };

    // Update recommendation
    updateRecConfig(extension.id, partialUpdate);
  });

  // Update settings recommendations
  updateSettings();
};

const isInitialized = shallowRef(false);
// Once recommendations has been returned from localStorage,
// we need to update to match what the user has installed
watch(
  () => recommendations.value,
  () => {
    if (isInitialized.value) {
      return;
    }

    isInitialized.value = true;
    getCurrentUserRecommendations();
  },
);

const useRecommendations = () => {
  const activeRecommendations = computed(() =>
    recommendations.value.filter((rec) => !rec.activated && !rec.ignored),
  );

  return {
    recommendations,
    activeRecommendations,
    updateRecConfig,
    updateSettings,
    getRecConfigById,
  };
};

export default useRecommendations;
