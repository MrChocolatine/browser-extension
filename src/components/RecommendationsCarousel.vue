<script lang="ts" setup>
import { NButton, NCard, NCarousel, NIcon, NImage } from 'naive-ui';
import ArrowForward from '~icons/mdi/arrow-right';
import ArrowBack from '~icons/mdi/arrow-left';

import { closePopup } from '@/helpers/closePopup';
import useRecommendations from '@/composables/useRecommendations';

const { recommendations } = useRecommendations();
</script>

<template>
  <h1 class="text-sm pb-1">Privacy Recommendations</h1>

  <div v-if="recommendations.length === 0">
    <p class="text-white text-lg">Sweet! You have taken action on all recommendations</p>
    <router-link class="hover:text-white" to="privacy-extensions"
      >See all Privacy Recommendations</router-link
    >
  </div>

  <div v-else>
    <n-carousel show-arrow>
      <template #arrow="{ prev, next }">
        <div class="custom-arrow">
          <n-icon class="arrow-icon" size="25" @click="prev"><ArrowBack /></n-icon>
          <n-icon class="arrow-icon" size="25" @click="next"><ArrowForward /></n-icon>
        </div>
      </template>
      <template #dots="{ total, currentIndex, to }">
        <ul class="custom-dots">
          <li
            v-for="index of total"
            :key="index"
            :class="{ ['is-active']: currentIndex === index - 1 }"
            @click="to(index - 1)"
          ></li>
        </ul>
      </template>

      <n-card v-for="(recommendation, index) in recommendations" :key="index" :bordered="false">
        <template #header>
          <div class="flex">
            <n-image
              v-if="recommendation.image"
              class="mr-4"
              width="20"
              :src="recommendation.image"
              object-fit="contain"
              preview-disabled
            />
            <h3>{{ recommendation.title }}</h3>
          </div>
        </template>

        <p>
          {{ recommendation.description }}
        </p>

        <div class="inline-flex items-center">
          <a
            v-if="recommendation.ctaURL"
            :href="recommendation.ctaURL"
            class="hover:text-white underline mr-4"
            @click="closePopup"
          >
            <n-button ghost>{{ recommendation.cta }}</n-button>
          </a>
          <router-link :to="recommendation.anchor" class="hover:text-white underline">
            Learn more
          </router-link>
        </div>
      </n-card>
    </n-carousel>

    <div class="text-right pt-2 mr-2">
      <router-link class="hover:text-white" to="privacy-recommendations">
        Show all ({{ recommendations.length }})
      </router-link>
    </div>
  </div>
</template>

<style>
.custom-arrow {
  display: flex;
  position: absolute;
  align-items: center;
  bottom: 9px;
  right: 11px;
}

.custom-arrow .n-icon {
  margin-left: 5px;
  cursor: pointer;
}

.custom-arrow .n-icon:hover {
  color: white;
}

.custom-arrow .n-icon:active {
  transform: scale(0.95);
  transform-origin: center;
}

.custom-dots {
  display: flex;
  margin: 0;
  padding: 0;
  position: absolute;
  bottom: 20px;
  left: 20px;
}

.custom-dots li {
  display: inline-block;
  width: 12px;
  height: 4px;
  margin: 0 3px;
  border-radius: 4px;
  background-color: rgba(255 255 255 / 40%);
  transition: width 0.3s, background-color 0.3s cubic-bezier(0.4, 0, 0.2, 1);
  cursor: pointer;
}

.custom-dots li.is-active {
  width: 40px;
  background: white;
}

.n-card > .n-card__content {
  min-height: 11em;
}
</style>
