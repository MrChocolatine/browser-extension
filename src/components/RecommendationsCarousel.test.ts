import { RouterLinkStub, shallowMount } from '@vue/test-utils';

import RecommendationsCarousel from '@/components/RecommendationsCarousel.vue';

import useRecommendations from '@/composables/useRecommendations/useRecommendations';

jest.mock('@/composables/useRecommendations/useRecommendations', () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe('RecommendationsCarousel', () => {
  it('should show a success message when no more recommendations are left', () => {
    (useRecommendations as jest.Mock).mockReturnValueOnce({ activeRecommendations: [] });
    const wrapper = shallowMount(RecommendationsCarousel, {
      global: { stubs: { RouterLink: RouterLinkStub } },
    });
    expect(wrapper.find('[data-test="success-message"]').exists()).toBe(true);
    expect(wrapper.text()).toMatch(/see all privacy recommendations$/i);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show a "Show all" link when there are recommendations', () => {
    (useRecommendations as jest.Mock).mockReturnValueOnce({ activeRecommendations: [{}, {}] });
    const wrapper = shallowMount(RecommendationsCarousel, {
      global: { stubs: { RouterLink: RouterLinkStub } },
    });
    expect(wrapper.find('[data-test="show-all-link"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="success-message"]').exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });
});
