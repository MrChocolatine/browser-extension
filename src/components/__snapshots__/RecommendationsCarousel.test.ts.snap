// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationsCarousel should show a success message when no more recommendations are left 1`] = `
<div
  data-v-app=""
>
  
  <h1
    class="font-semibold text-xl mb-2 text-gray-200"
  >
    Privacy recommendations
  </h1>
  <div
    class="n-card n-card--bordered space-y-4"
    style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border-radius: 3px; --n-color: #fff; --n-color-modal: #fff; --n-color-popover: #fff; --n-color-target: #18a058; --n-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-action-color: rgb(250, 250, 252); --n-title-text-color: rgb(31, 34, 37); --n-title-font-weight: 500; --n-close-color: rgba(122, 122, 122, 1); --n-close-color-hover: rgba(89, 89, 89, 1); --n-close-color-pressed: rgba(149, 149, 149, 1); --n-border-color: rgb(239, 239, 245); --n-box-shadow: 0 1px 2px -2px rgba(0, 0, 0, .08), 0 3px 6px 0 rgba(0, 0, 0, .06), 0 5px 12px 4px rgba(0, 0, 0, .04); --n-padding-top: 19px; --n-padding-bottom: 20px; --n-padding-left: 24px; --n-font-size: 14px; --n-title-font-size: 18px; --n-close-size: 18px;"
  >
    <!---->
    <!---->
    <div
      class="n-card__content"
      role="none"
    >
      <span
        data-test="success-message"
      >
        <i
          class="mr-2 n-icon"
          role="img"
          style="--n-bezier: cubic-bezier(.4, 0, .2, 1); font-size: 25px;"
        >
          <svg
            class="text-success"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
        </i>
         Sweet! You have taken action on all recommendations.
      </span>
    </div>
    <!---->
    <!---->
  </div>
  <div
    class="text-right pt-2 mr-2"
  >
    <a
      class="hover:text-white"
    >
       Show all 
    </a>
  </div>
  
</div>
`;
